### PR TITLE
Shared notes

### LINKED ISSUE #52: JS Interactions: shared users

### LINKED ISSUE #53: JS Interactions: readonly and public notes

### LINKED ISSUE #62: Delete note

### PR DESCRIPTION
**Pull request by Blatoy**
Close #52 
Close #53 
Close #62

### CODE REVIEW

File: marcdown_project/frontend/src/components/Modal.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    hide() {
+        this.setState({
+            displayed: false
+        });
+    }
+
+    hideOnClick(e) {
+        if (e.target.classList.contains("modal-background")) {
+            this.setState({
```
Comment (g-wizzy) [inline marcdown_project/frontend/src/components/Modal.js]: DRY

File: marcdown_project/frontend/src/pages/Editor.js [hunk 2]
```diff
@@ -1,108 +1,220 @@
...
+        if (this.state.existsInDatabase) {
+            const patchText = dmp.patch_toText(dmp.patch_make(this.state.previousSavedText, this.state.input));
+
+            query(`/api/note/${this.state.noteId}/`, "PATCH", {
+                diff: patchText
+            }).then((result) => { });
+            this.state.previousSavedText = this.state.input;
+        } else {
+            // Create a new note
+            query(`/api/note/`, "POST", {
```
Comment (g-wizzy) [inline marcdown_project/frontend/src/pages/Editor.js:50]: Why not send actual values ?

File: marcdown_project/frontend/src/pages/Editor.js [hunk 3]
```diff
@@ -1,108 +1,220 @@
...
-    } else {
-      query(`/api/note/${id}`).then((result) => {
-        if (result.id === undefined) {
-          this.setState({ defaultInput: "# Could not load this note\n\n**Error detail**: " + result.detail + "\n\nMake sure you have the permission to read this note\n\nEdit this note to create a new one" });
+        query(`/api/note/${this.state.noteId}/`, "PUT", {
+            sharedWith: newSharedArray
+        }).then((result) => {
+        });
+
+        this.setState({ sharedWith: newSharedArray });
```
Comment (g-wizzy) [inline marcdown_project/frontend/src/pages/Editor.js:101]: Shouldn't this be done when the API confirms the new sharers ?

File: marcdown_project/marcdown/models/note.py [hunk 4]
```diff
@@ -21,6 +21,9 @@ class Note(models.Model):
     def __str__(self):
         return self.title
     
+    def is_owner(self, user):
+        return hasattr(user, 'profile') and self.owner == user.profile
```
Comment (g-wizzy) [inline marcdown_project/marcdown/models/note.py]: Perhaps `user.is_authenticated` would be more relevant and readable than the `hasattr`

File: marcdown_project/frontend/src/pages/Editor.js (hunk 3 above)
Comment (Blatoy) [reply marcdown_project/frontend/src/pages/Editor.js:101]: This is done so the user instantly disappear when the user click. This should only fail if there is an issue with the API

File: marcdown_project/marcdown/api_views/notes.py [hunk 5]
```diff
@@ -17,8 +17,11 @@ def retrieve(self, request, pk=None):
         Gets the note corresponding to the given id,
         and responds with its description and content
         '''
+        user = request.user
+
         queryset = Note.objects.all()
         note = get_object_or_404(queryset, id=pk)
+        note.is_owner = note.is_owner(user)
```
Comment (g-wizzy) [inline marcdown_project/marcdown/api_views/notes.py:24]: Overriding the model's `is_owner` method is not a good idea

**Review by g-wizzy — CHANGES_REQUESTED** [review]

**Review by g-wizzy — APPROVED** [review]

**Pull request merged**